### PR TITLE
Fix relative metadata paths in SequentialCompressionWriter

### LINKED ISSUE #557: Recording with compression creates non-standard relative filepaths in metadata
## Description
Running a bag record operation with compression enabled generates relative file paths that include the directory within the metadata. If compression is disabled, the relative file paths do not include the bag directory.

Ex: If recording a bag `bar` in the `foo` directory:

Filesystem View:
```
/../../foo/
       | bar/
       |  |  bar_0.sql.zstd
       |  |  bar_1.sql.zstd
       |  |  bar_2.sql.zstd
```

Relative file paths in metadata for compressed case:
- bar/bar_0.sql.zstd
- bar/bar_1.sql.zstd
- bar/bar_2.sql.zstd

Relative file paths in metadata for uncompressed case:
- bar_0.sql
- bar_1.sql
- bar_2.sql

### Expected Behavior
Consistent relative file path format within `metadata.yaml` across compression modes

### Actual Behavior
Relative filepath format within `metadata.yaml` differs between compression modes

## To Reproduce
1. Record uncompressed bag
2. Record compressed bag (use file compression, haven't checked this with message-based compression)
3. Compare `metadata.yaml` files between bags. The key of interest is `relative_file_paths`

## System (please complete the following information)
 - OS: Ubuntu Bionic
 - ROS 2 Distro: Rolling
 - Version: master branch (issue discovered on personal fork, however the compression and bag-writing code has not been modified, so this issue should be visible on master)

### PR DESCRIPTION
**Pull request by emersonknapp**
Fixes #557
Related to #612 

This makes the SequentialCompressionWriter write in the bag metadata with correct relative paths in the V4 format.

* Use common init_metadata logic to get the correct relative path there
* In the file compression, be much more careful about what the paths truly mean, given that the metadata requires a path relative to the bag, but the compressor just needs to find and compress a file on the filesystem
* Factor out some common test logic so the new test can be a bit more readable

### CODE REVIEW

File: rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp [hunk 1]
```diff
@@ -244,23 +243,23 @@ void SequentialCompressionWriter::compress_file(
...
+          "Failed to find path to uncompressed bag: \"" << file_relative_to_pwd.string() <<
             "\"; this shouldn't happen.");
       }
     }
 
-    if (!rcpputils::fs::remove(to_compress)) {
+    if (!rcpputils::fs::remove(file_relative_to_pwd)) {
       ROSBAG2_COMPRESSION_LOG_ERROR_STREAM(
-        "Failed to remove uncompressed bag: \"" << to_compress.string() << "\"");
+        "Failed to remove uncompressed bag: \"" << file_relative_to_pwd.string() << "\"");
```
Comment (zmichaels11) [inline rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp:258]: This logs an error without throwing.
Should this behavior be reported in the documentation for the function?

File: rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp [hunk 2]
```diff
@@ -270,13 +269,15 @@ void SequentialCompressionWriter::split_bagfile()
   std::lock_guard<std::recursive_mutex> lock(storage_mutex_);
   std::lock_guard<std::mutex> compressor_lock(compressor_queue_mutex_);
 
-  switch_to_next_storage();
+  // Grab last file before calling common splitting logic, which pushes the new filename
+  std::string last_file = metadata_.relative_file_paths.back();
```
Comment (zmichaels11) [inline rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp:273]: `const auto last_file`?

**Review by zmichaels11 — APPROVED** [review]

**Review by prajakta-gokhale — APPROVED** [review]

**Comment by emersonknapp** [comment]
Gist: https://gist.githubusercontent.com/emersonknapp/48d4065acf325d5ca483591e2588d687/raw/cc4ce0d6c338621125898050f630578454945852/ros2.repos
BUILD args: --packages-up-to rosbag2_compression rosbag2_tests
TEST args: --packages-select rosbag2_compression rosbag2_tests
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/7513
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13493/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8400/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11215/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13558/)

File: rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp [hunk 3]
```diff
@@ -240,13 +236,16 @@ void SequentialCompressionWriter::remove_topic(
...
-  const std::string & file)
+  const std::string & file_relative_to_bag)
 {
-  ROSBAG2_COMPRESSION_LOG_DEBUG("Compressing file: %s", file.c_str());
-  const auto to_compress = rcpputils::fs::path{file};
+  using rcpputils::fs::path;
 
-  if (to_compress.exists() && to_compress.file_size() > 0u) {
-    const auto compressed_uri = compressor.compress_uri(to_compress.string());
+  auto file_relative_to_pwd = path(base_folder_) / file_relative_to_bag;
```
Comment (zmichaels11) [inline rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp:243]: Can `file_relative_to_pwd` be const?

**Review by zmichaels11 — APPROVED** [review]

**Comment by emersonknapp** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13525/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8426/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11249/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13589/)

**Pull request merged**